### ISSUE TITLE
A schema where property value is limited to another zod schema of any type

### ISSUE THREAD

**Issue by rudfoss** [colinhacks/zod#1016]
Hi, first of all. Thanks for an awesome and intuitive library! I've been using it for multiple projects and bragging about it to other developers :D

I'm wondering if there is a way to declare a schema where a single property value is itself a `zod` schema. I'm trying to define some rules for a specific object type, but leave the "value" validation up to the object itself. Something like this:

```typescript
const FieldDefinition = z.object({
  title: z.string(),
  value: // type of any valid zod schema
})
type FieldDefinitionType = z.infer<typeof FieldDefinition>
```

Then in some other part of my program I want to be able to do something like this:

```typescript
const TextField: FieldDefinitionType = {
  title: "Text field",
  value: z.string().length(128)
}
```

Basically limiting the `value` property to only be another `zod` schema. How would I type that up in the `FieldDefinition` schema?

Edit: To be clear. The value should not be limited to `z.string()` or `z.number()`, any valid zod schema should be allowed.

**Comment by rudfoss** [comment]
Right now I've built a small, custom schema to solve the problem just for my particular use case, but maybe there is a better way?

```typescript
export const ZodSchema = z.object({
  safeParse: z
    .function()
    .args(z.unknown(), z.any().optional())
    .returns(
      z.object({
        success: z.boolean(),
        data: z.instanceof(ZodError).optional(),
        error: z.any().optional()
      })
    )
})
```

**Comment by FlorianWendelborn** [comment]
You could use a factory for the schema you want. For example:

```ts
import { z } from 'zod'

const commonSchema = z.object({
    some: z.literal('thing')
})

type Common = z.output<typeof commonSchema>

const getFieldSchema = <VALUE extends unknown>(valueSchema: z.ZodSchema<VALUE>) => commonSchema.extend({
    value: valueSchema
})

const textFieldSchema = getFieldSchema(z.string())
type TextField = z.output<typeof textFieldSchema>

const multiSelectFieldSchema = getFieldSchema(z.array(z.string()))
type MultiSelectField = z.output<typeof multiSelectFieldSchema>
```

[TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgFBkDGEAdgM7zUgi0DKlAFgKYgCGcAXhQA6CACMAVl0owAFAjJwlcOri4AuEQBtgMLlF5bZ+GB2A0A5vgCUZNLbIwAnmC5wAwrhY1BIiAFcYMECAHmdXCEw4Jm92bj4APgpqengLLhgAMWAuLWI4nn4hEIA1AEEAGQBVAFE4LgAPPRpiOjh-GgBrGggAdxoE2QA3Q38uAr5NZGEALRIJ3lLK2oTrQQTorzZOQuFG5uJ5RWURrTHNU7GFuwcUhjg9Juzc-J2+X3SsnLyF2WmGKDmCyyay2cJuAAq+2eeV80wCQVC4MiD2h31e8V4SSotHuIH8WhgwFYuWkXxeCw+GRhGMKf2EvCgBic9IBQJBoMcLjcAFkCUSSVoyTS4aJAsEYGFuSj8YTiaSZDSFgkgA)

**Comment by rudfoss** [comment]
> You could use a factory for the schema you want. For example:
> 
> ```ts
> import { z } from 'zod'
> 
> const commonSchema = z.object({
>     some: z.literal('thing')
> })
> 
> type Common = z.output<typeof commonSchema>
> 
> const createFieldSchema = <VALUE extends unknown>(valueSchema: z.ZodSchema<VALUE>) => commonSchema.extend({
>     value: valueSchema
> })
> 
> const textFieldSchema = createFieldSchema(z.string())
> type TextField = z.output<typeof textFieldSchema>
> 
> const multiSelectFieldSchema = createFieldSchema(z.array(z.string()))
> type MultiSelectField = z.output<typeof multiSelectFieldSchema>
> ```
> 
> [TypeScript Playground](https://www.typescriptlang.org/play?ssl=9&ssc=1&pln=10&pc=1#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgFBkDGEAdgM7zUgi0DKlAFgKYgCGcAXhQA6CACMAVl0owAFAjJwlcOri4AuEQBtgMLlF5bZ+GB2A0A5vgCUZNLbIwAnmC5wAwrhY1BIiAFcYMECAHmdXCEw4Jm92bj4APgpqekYoLl49ADFgLi1iOJ5+IRCANQBBABkAVQBROC4ADz0aYjo4fxoAaxoIAHcaBNkAN0N-LkK+TWRhAC0SSd4yqrqE60EE6K82TiLhJpbieUVlUa1xzTPxxbsHFIY4PWacvILdvl9KdMyuF-zF2QzBhQcwWWTWWzhNwAFQOf2IvhmASCoShkUecNy-3evCSVFoDxA-i0MGArDy0hg8MWn2+2Sxb3ivEBwl4UAMThZwNB4IhjhcbgAssTSeStJT4YjRIFgjAwgL0USSWSKTJqTiEkA)

That's a great idea! It will also allow me to apply restrictions and do things like:

```typescript
const createFieldSchema = <T extends unknown>(baseField: FieldDefinitionType, valueSchema: z.ZodSchema<T>) => ...
```

Thanks for the help!
